// wilderness/lib/core/shape/props.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var animationDefaults = {
  delay: 0,
  duration: 250
};

var animationProps = ['delay', 'duration', 'easing', 'finish', 'motionPath', 'name', 'start', 'update'];

var manipulationProps = ['moveIndex', 'offset', 'reverse', 'scale'];

var shapeProps = ['cx', 'cy', 'd', 'height', 'points', 'r', 'rx', 'ry', 'shapes', 'type', 'width', 'x', 'x1', 'x2', 'y', 'y1', 'y2'];

var stylePropAttrMap = {
  fill: 'fill',
  fillOpactity: 'fill-opacity',
  fillRule: 'fill-rule',
  stroke: 'stroke',
  strokeDasharray: 'stroke-dasharray',
  strokeDashoffset: 'stroke-dashoffset',
  strokeLinecap: 'stroke-linecap',
  strokeLinejoin: 'stroke-linejoin',
  strokeOpactity: 'stroke-opacity',
  strokeWidth: 'stroke-width',
  vectorEffect: 'vector-effect'
};

var styleProps = Object.keys(stylePropAttrMap);

exports.animationDefaults = animationDefaults;
exports.animationProps = animationProps;
exports.manipulationProps = manipulationProps;
exports.shapeProps = shapeProps;
exports.stylePropAttrMap = stylePropAttrMap;
exports.styleProps = styleProps;
